core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -455,7 +455,23 @@ public IndexMetaData index(String index) {
     }
 
     public IndexMetaData index(Index index) {
-        return index(index.getName());
+        IndexMetaData metaData = index(index.getName());
+        if (metaData != null && metaData.getIndexUUID().equals(index.getUUID())) {
+            return metaData;
+        }
+        return null;
+    }
+
+    /**
+     * Returns the {@link IndexMetaData} for this index.
+     * @throws IndexNotFoundException if no metadata for this index is found
+     */
+    public IndexMetaData getIndexSafe(Index index) {
+        IndexMetaData meta = index(index);
+        if (meta == null) {
+            throw new IndexNotFoundException(index);
```

Comment:
Maybe distinguish whether no index with that name exists or whether index with that name has different UUID and have a different error message? Could be useful for debugging purposes.